src/sql/src/query_model/lowering.rs

```diff
@@ -0,0 +1,222 @@
+    fn new(model: &'a Model) -> Self {
+        Self { model }
+    }
+
+    /// Applies the given box identified by its ID to the given outer relation.
+    fn apply(
+        &mut self,
+        box_id: BoxId,
+        get_outer: expr::MirRelationExpr,
+        outer_column_map: &ColumnMap,
+    ) -> expr::MirRelationExpr {
+        use expr::MirRelationExpr as SR;
+        let the_box = self.model.get_box(box_id).borrow();
+        match &the_box.box_type {
+            BoxType::Values(values) => {
+                let identity = SR::constant(vec![vec![]], RelationType::new(vec![]));
+                // TODO(asenac) lower actual values
+                assert!(values.rows.len() == 1 && the_box.columns.len() == 0);
+                get_outer.product(identity)
+            }
+            BoxType::Select(select) => {
+                // A Select box combines three operations, join, filter and project,
+                // in that order.
+
+                // 1) Lower the join component of the Select box.
+                // TODO(asenac) We could join first all non-correlated quantifiers
+                // and then apply the correlated ones one by one in order of dependency
+                // on top the join built so far, adding the predicates as soon as their
+                // dependencies are satisfied.
+                let correlation_info = the_box.correlation_info(&self.model);
+                if !correlation_info.is_empty() {
+                    panic!("correlated joins are not supported yet");
+                }
+
+                let outer_arity = get_outer.arity();
+                let (mut input, column_map) =
+                    self.lower_join(get_outer, outer_column_map, &the_box.quantifiers);
+                let input_arity = input.arity();
+
+                // 2) Lower the filter component.
+                if !select.predicates.is_empty() {
+                    input = input.filter(
+                        select
+                            .predicates
+                            .iter()
+                            .map(|p| Self::lower_expression(p, &column_map)),
+                    );
+                }
+
+                // 3) Lower the project component.
+                if !the_box.columns.is_empty() {
+                    input = input.map(
+                        the_box
+                            .columns
+                            .iter()
+                            .map(|c| Self::lower_expression(&c.expr, &column_map))
+                            .collect_vec(),
+                    );
+                }
+
+                // Project the outer columns plus the ones in the projection of
+                // this select box
+                input.project(
+                    (0..outer_arity)
+                        .chain(input_arity..input_arity + the_box.columns.len())
+                        .collect_vec(),
+                )
+            }
+            _ => panic!(),
+        }
```

Comment:
What are we doing with the `distinct` field of `the_box`? If `distinct = DistinctOperation::Enforce` we need to wrap the result in a `.distinct()`.

Reply:
It's ignored, since it is untestable at the moment, since we don't lower `HirRelationExpr::Distinct` yet.

Reply:
I could an assert for those things that are not supported yet.

Reply:
The problem is that I am not sure what mechanism we use to keep track outstanding work items. We have the tracking issue but the tasks there are not really up-to-date. We need to decide how we are going do to it consistently and then stick that decision.